src/token.rs

```diff
@@ -1,12 +1,73 @@
+use lazy_static::__Deref;
+
 /// This struct contains a token, an expiry, and an access scope.
 pub struct Token {
     // this field contains the JWT and the expiry thereof. They are in the same Option because if
     // one of them is `Some`, we require that the other be `Some` as well.
-    token: Option<(String, u64)>,
+    token: tokio::sync::RwLock<Option<(String, u64)>>,
     // store the access scope for later use if we need to refresh the token
     access_scope: String,
 }
 
+/// Trait that refreshes a token when it is expired
+#[async_trait::async_trait]
+pub trait RefreshableToken: Sync {
+    /// Getter for the token
+    async fn get_inner_token(&self) -> Option<(String, u64)>;
+    /// Updates the token
+    async fn update_inner_token(&self, token: (String, u64)) -> crate::Result<()>;
+    /// Getter for the scope
+    async fn get_scope(&self) -> &str;
+
+    /// Fetches then returns an unexpired token
+    /// Updates the token when it is expired
+    async fn get(&self, client: &reqwest::Client) -> crate::Result<String> {
+        match self.get_inner_token().await {
+            Some((token, exp)) if exp > now() => Ok(token),
+            _ => {
+                let scope = self.get_scope().await;
+                let token = Self::fetch_token(client, scope).await?;
+                self.update_inner_token(token).await?;
+
+                Ok(self.get_inner_token().await.unwrap().0)
+            }
+        }
+    }
+
+    /// Fetches and returns the token using the service account
+    async fn fetch_token(client: &reqwest::Client, scope: &str) -> crate::Result<(String, u64)> {
+        let now = now();
+        let exp = now + 3600;
+
+        let claims = Claims {
+            iss: crate::SERVICE_ACCOUNT.client_email.clone(),
+            scope: scope.into(),
+            aud: "https://www.googleapis.com/oauth2/v4/token".to_string(),
+            exp,
+            iat: now,
+        };
+        let header = jsonwebtoken::Header {
+            alg: jsonwebtoken::Algorithm::RS256,
+            ..Default::default()
+        };
+        let private_key_bytes = crate::SERVICE_ACCOUNT.private_key.as_bytes();
```

Comment:
Should we use the global service account private key here? This might not be useful for most user provided implementations of `RefreshableToken`.

Reply:
So this is a default implementation of how the token is retrieved. I envisioned this method be overridden however a custom implementation is needed. But you bring up a related point to whether this implementation should be somewhere else

Reply:
I think refactoring the token itself as a `ServiceAccountToken` or `SAToken` might be a better fit here. What do you think @ThouCheese